Tekton controller panic at `github.com/tektoncd/pipeline/pkg/reconciler/taskrun.(*Reconciler).durationAndCountMetrics`

# Expected Behavior

The Tekton controller should not panic.

# Actual Behavior

The Tekton controller logs this ...:

```json
{
    "level": "error",
    "ts": "2022-07-14T15:00:55.806Z",
    "logger": "tekton-pipelines-controller",
    "caller": "taskrun/taskrun.go:196",
    "msg": "Error getting TaskRun taskrun1 when updating metrics: %!w(*errors.StatusError=&{{{ } {   <nil>} Failure taskrun.tekton.dev \"taskrun1\" not found NotFound 0xc00037c7e0 404}})",
    "commit": "0d8b26a",
    "knative.dev/controller": "github.com.tektoncd.pipeline.pkg.reconciler.taskrun.Reconciler",
    "knative.dev/kind": "tekton.dev.TaskRun",
    "knative.dev/traceid": "59f10992-1478-49b7-b2af-63d6cea0cab6",
    "knative.dev/key": "namespace1/taskrun1",
    "stacktrace": "github.com/tektoncd/pipeline/pkg/reconciler/taskrun.(*Reconciler).durationAndCountMetrics\n\tgithub.com/tektoncd/pipeline/pkg/reconciler/taskrun/taskrun.go:196\ngithub.com/tektoncd/pipeline/pkg/reconciler/taskrun.(*Reconciler).reconcile\n\tgithub.com/tektoncd/pipeline/pkg/reconciler/taskrun/taskrun.go:451\ngithub.com/tektoncd/pipeline/pkg/reconciler/taskrun.(*Reconciler).ReconcileKind\n\tgithub.com/tektoncd/pipeline/pkg/reconciler/taskrun/taskrun.go:173\ngithub.com/tektoncd/pipeline/pkg/client/injection/reconciler/pipeline/v1beta1/taskrun.(*reconcilerImpl).Reconcile\n\tgithub.com/tektoncd/pipeline/pkg/client/injection/reconciler/pipeline/v1beta1/taskrun/reconciler.go:235\nknative.dev/pkg/controller.(*Impl).processNextWorkItem\n\tknative.dev/pkg@v0.0.0-20220329144915-0a1ec2e0d46c/controller/controller.go:542\nknative.dev/pkg/controller.(*Impl).RunContext.func3\n\tknative.dev/pkg@v0.0.0-20220329144915-0a1ec2e0d46c/controller/controller.go:491"
}
```

... and then panics:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1974442]
goroutine 254 [running]:
github.com/tektoncd/pipeline/pkg/reconciler/taskrun.(*Reconciler).durationAndCountMetrics(0xc00004c2c0, {0x2230f30?, 0xc000e77e00?}, 0xc000163080)
	github.com/tektoncd/pipeline/pkg/reconciler/taskrun/taskrun.go:198 +0x1a2
github.com/tektoncd/pipeline/pkg/reconciler/taskrun.(*Reconciler).reconcile(0xc00004c2c0, {0x2230f30, 0xc000e77e00}, 0xc000163080, 0x0?)
	github.com/tektoncd/pipeline/pkg/reconciler/taskrun/taskrun.go:451 +0x148d
github.com/tektoncd/pipeline/pkg/reconciler/taskrun.(*Reconciler).ReconcileKind(0xc00004c2c0, {0x2230f30, 0xc000e77c80}, 0xc000163080)
	github.com/tektoncd/pipeline/pkg/reconciler/taskrun/taskrun.go:173 +0x988
github.com/tektoncd/pipeline/pkg/client/injection/reconciler/pipeline/v1beta1/taskrun.(*reconcilerImpl).Reconcile(0xc0002b99a0, {0x2230f30, 0xc000e77b00}, {0xc000b7b540, 0x37})
	github.com/tektoncd/pipeline/pkg/client/injection/reconciler/pipeline/v1beta1/taskrun/reconciler.go:235 +0x582
knative.dev/pkg/controller.(*Impl).processNextWorkItem(0xc0003e31a0)
	knative.dev/pkg@v0.0.0-20220329144915-0a1ec2e0d46c/controller/controller.go:542 +0x58d
knative.dev/pkg/controller.(*Impl).RunContext.func3()
	knative.dev/pkg@v0.0.0-20220329144915-0a1ec2e0d46c/controller/controller.go:491 +0x68
created by knative.dev/pkg/controller.(*Impl).RunContext
	knative.dev/pkg@v0.0.0-20220329144915-0a1ec2e0d46c/controller/controller.go:489 +0x312
```

In our scenario there's another controller that creates TaskRuns, watches them and deletes them once they are finished (`Succeeded = true`). `pkg/reconciler/taskrun.(*Reconciler).durationAndCountMetrics` is in a race condition with it.

It may be relevant that in out scenario the whole K8s namespace gets deleted, where the deletion of the TaskRun (and any other objects in such namespace) is a cascaded deletion.

# Steps to Reproduce the Problem

1. Run an additional program that watches for changes of a TaskRun and deletes it once it transitioned to `Succeeded == true`.
2. Create a TaskRun and wait until it succeeded.
3. Maybe the Tekton controller panicked, depending on the position in the race with the additional program.

# Additional Info

- Kubernetes version:

  **Output of `kubectl version`:**

  ```
  Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.11", GitCommit:"5824e3251d294d324320db85bf63a53eb0767af2", GitTreeState:"clean", BuildDate:"2022-06-16T05:33:55Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
  ```

- Tekton Pipeline version: `v0.36.0`

<!-- Any other additional information -->

--- a/pkg/reconciler/pipelinerun/pipelinerun.go
+++ b/pkg/reconciler/pipelinerun/pipelinerun.go
@@ -247,9 +247,13 @@ func (c *Reconciler) durationAndCountMetrics(ctx context.Context, pr *v1beta1.Pi
 	if pr.IsDone() {
 		// We get latest pipelinerun cr already to avoid recount
 		newPr, err := c.pipelineRunLister.PipelineRuns(pr.Namespace).Get(pr.Name)
-		if err != nil {
+		if err != nil && !kerrors.IsNotFound(err) {
 			logger.Errorf("Error getting PipelineRun %s when updating metrics: %w", pr.Name, err)
+			return
+		} else if kerrors.IsNotFound(err) {
+			logger.Debugf("Pipelinerun %s not found when updating metrics: %w", pr.Name, err)
 		}
+
 		before := newPr.Status.GetCondition(apis.ConditionSucceeded)
 		go func(metrics *pipelinerunmetrics.Recorder) {
 			err := metrics.DurationAndCount(pr, before)

--- a/pkg/reconciler/taskrun/taskrun.go
+++ b/pkg/reconciler/taskrun/taskrun.go
@@ -218,9 +218,13 @@ func (c *Reconciler) durationAndCountMetrics(ctx context.Context, tr *v1beta1.Ta
 	logger := logging.FromContext(ctx)
 	if tr.IsDone() {
 		newTr, err := c.taskRunLister.TaskRuns(tr.Namespace).Get(tr.Name)
-		if err != nil {
+		if err != nil && !k8serrors.IsNotFound(err) {
 			logger.Errorf("Error getting TaskRun %s when updating metrics: %w", tr.Name, err)
+			return
+		} else if k8serrors.IsNotFound(err) {
+			logger.Debugf("TaskRun %s not found when updating metrics: %w", tr.Name, err)
 		}
+
 		before := newTr.Status.GetCondition(apis.ConditionSucceeded)
 		go func(metrics *taskrunmetrics.Recorder) {
 			if err := metrics.DurationAndCount(ctx, tr, before); err != nil {

